WebUI: Settings & 4 different ways to override

petemill opened the pull request:
Two main areas to this PR: The Settings page, and architectural changes to the way we modify chromium WebUIs.


## WebUI Settings
- A fixed page sub-navigation (on wider windows)
- Inject a get-started section
- Brave-styles across more material components
- Fix some issues with Brave's added sections (e.g. Extensions) showing up during all searches

Fix https://github.com/brave/brave-browser/issues/958
Fix https://github.com/brave/brave-browser/issues/960
Fix https://github.com/brave/brave-browser/issues/955
Address https://github.com/brave/brave-browser/issues/26 (this is more of an umbrella issue)
Fix https://github.com/brave/brave-browser/issues/964
Fix https://github.com/brave/brave-browser/issues/3444
Fix https://github.com/brave/brave-browser/issues/3826


## WebUI General overriding

1. Add a behavior to an existing Polymer JS component. This is similar to injecting a superclass which will receive the same lifecycle method calls. It could also be used to add properties and methods to the component, usable from the component's html template.

2. Modify an existing Polymer HTML template. When the component is registered, we provide the template DOM element which can be modified before it is cloned for use as content and attached to the document.

3. Inject style to any existing Polymer HTML component. By naming a module 'brave-override-style-my-module', the <style> element from that module will be taken and injected in to the existing 'my-module' component.

4. Replace icons in an existing iconset. This takes an icon from one existing iconset and replaces the content of an icon in another existing iconset. Can be used to replace a chromium icon with a brave icon.

Whilst all this can be achieved via .patch files at sync / build time, this new methodology:
- Keeps all the modifications functional. The advantage here is realized in the 100% inevitable and likely-common scenarios where the chromium webui implementations and underlying technology changes (e.g. html-modules vs es-modules, or simple module refactors). For some scenarios, we may be able to simply modify the central manipulation library (polymer_overrides.js). In more complicated scenarios, we will at least have all modifications functionally and individually labelled, making 'rebasing' potentially more straightforward than a .patch file with mixed intentions.
- Maintains performance since these modifications are only performed once, whilst the component is being registered via the Polymer(component) function.

Debugging:
- Display (in verbose console log) how long it took to modify each Polymer template
- Option to see timing of module and module-override registration (via hard-coded debug variable)

----

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:
- Open chrome://settings
- should look like:
![image](https://user-images.githubusercontent.com/741836/54582033-7402b000-49cc-11e9-886b-19c8de3f959f.png)
- also see test plan in:
    - https://github.com/brave/brave-browser/issues/958
    - https://github.com/brave/brave-browser/issues/960
    - https://github.com/brave/brave-browser/issues/955
    - https://github.com/brave/brave-browser/issues/3444

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

petemill commented:
I'm done with functionality for this PR. Just dealing with a slightly complicated rebase now.

petemill commented:
Some pure refactoring follow-ups I'd like to do for WebUIs after this:

- Move some of the other patching to this style of functional modification
- Split up `settings_overrides.js` in to functional files instead of a single large file.
- Consider using webpack to generate certain JS files (such as `settings_overrides.js`)
- More brave-styles for material components

bsclifton commented:
I rebased this so I can build- after a completely fresh Release build (deleted src/out, no sccache), I see the following (notice the weird border on the right):
<img width="373" alt="Screen Shot 2019-03-19 at 12 43 01 PM" src="https://user-images.githubusercontent.com/4733304/54636917-a7812100-4a44-11e9-83e2-4f6b61db6a24.png">

bsclifton commented:
I think the padding/margin under `About Brave` is a bit excessive. From the original mock-up:
<img width="285" alt="Screen Shot 2019-03-19 at 12 47 57 PM" src="https://user-images.githubusercontent.com/4733304/54637184-3ee67400-4a45-11e9-8216-3171136e56bc.png">

bsclifton commented:
I remember there were some difficulties- but wanted to check. Shouldn't there be icons next to each of the titles?

mock-up (shown in https://github.com/brave/brave-browser/issues/955 and https://github.com/brave/brave-browser/issues/964):
<img width="463" alt="Screen Shot 2019-03-19 at 12 48 49 PM" src="https://user-images.githubusercontent.com/4733304/54637283-840aa600-4a45-11e9-8e18-aa2f7a04d877.png">

implementation:
<img width="179" alt="Screen Shot 2019-03-19 at 12 45 57 PM" src="https://user-images.githubusercontent.com/4733304/54637355-b2888100-4a45-11e9-9fca-6f1cb4bc289a.png">

bsclifton commented:
@srirambv already captured with https://github.com/brave/brave-browser/issues/3633 - but the focus ring (aria?) looks odd. Not sure if border radius can be added? We don't have to solve this now of course
<img width="164" alt="Screen Shot 2019-03-19 at 1 06 03 PM" src="https://user-images.githubusercontent.com/4733304/54638404-35aad680-4a48-11e9-9b11-28aec0785740.png">

bsclifton commented:
This new section needs an entry on the sidebar nav (and icon):
<img width="213" alt="Screen Shot 2019-03-19 at 1 13 05 PM" src="https://user-images.githubusercontent.com/4733304/54638669-cbdefc80-4a48-11e9-9a92-1e9590eb9712.png">

bsclifton commented:
Not sure if it's because I'm on a fresh profile? but trying to visit any of the autofill options leaves me at a blank page
<img width="678" alt="Screen Shot 2019-03-19 at 1 18 20 PM" src="https://user-images.githubusercontent.com/4733304/54638986-8ff86700-4a49-11e9-9f40-b5f6522fc4b1.png">

Clicking goes to:
<img width="1023" alt="Screen Shot 2019-03-19 at 1 19 10 PM" src="https://user-images.githubusercontent.com/4733304/54639014-9edf1980-4a49-11e9-81a1-2594db2249f2.png">

Expected to see something like:
<img width="957" alt="Screen Shot 2019-03-19 at 1 19 35 PM" src="https://user-images.githubusercontent.com/4733304/54639047-b3231680-4a49-11e9-906c-a6968bc1c496.png">

bsclifton commented:
@petemill I updated all the issues closed and added the appropriate milestone/labels/test plans 😄 Please just double check those

Functionally - I called out everything I found when reviewing above 😄 Great job on this!

Going to continue reviewing / looking at the code

bsclifton commented on browser/resources/settings/brave_settings_overrides.js:
```diff
@@ -1,9 +1,317 @@
+      // then section is no longer 'selected'.
+      // TODO(petemill): If this wasn't a chromium module, we'd simply add a handler
+      // for scrolling away, or have the menu change selection as we scroll.
+      stopObservingScroll()
```
curious about this code... let's say I click `Get Started`. Sidebar properly shows it selected (orange). As I scroll down, it does "go out of scope" and un-select itself. But I'm not seeing `Extensions`, `Sync`, `Appearance`, etc select themselves. Is the highlighted section supposed to be shown? (also, if so- maybe we can call out in test plan)

petemill replied:
@bsclifton no support for scrolling as you move. Just kept the current chromium highlighting behavior and removed it when scrolling away so it's not confusing. As the comment says, if we want to add more custom code then we can add a feature to have the menu change selection as we scroll, but I'm not sure I'd be in favor of that as it gets complicated when there are multiple items on the screen, or the scroll never gets that far.

For now, the selection on clicking an item is a nice feedback to the menu interaction.

> ```
> // TODO(petemill): If this wasn't a chromium module, we'd simply add a handler
> // for scrolling away, or have the menu change selection as we scroll.
> ```

bsclifton commented on browser/resources/settings/brave_settings_overrides.js:
```diff
@@ -1,9 +1,317 @@
+
+// Templates
+BravePatching.RegisterPolymerTemplateModifications({
+  'settings-ui': (templateContent) => {
```
for other reviewers: the keys here (ex: `settings-ui`) are the names of components on the page 😄

petemill commented:
> @srirambv already captured with [brave/brave-browser#3633](https://github.com/brave/brave-browser/issues/3633) - but the focus ring (aria?) looks odd. Not sure if border radius can be added? We don't have to solve this now of course
> <img alt="Screen Shot 2019-03-19 at 1 06 03 PM" width="164" src="https://user-images.githubusercontent.com/4733304/54638404-35aad680-4a48-11e9-9b11-28aec0785740.png">

**Button style was changed before this PR, so can be addressed separately.**

> I remember there were some difficulties- but wanted to check.
> 
> _Shouldn't there be icons next to each of the titles?_
> 
> mock-up (shown in [brave/brave-browser#955](https://github.com/brave/brave-browser/issues/955) and [brave/brave-browser#964](https://github.com/brave/brave-browser/issues/964)):
> <img alt="Screen Shot 2019-03-19 at 12 48 49 PM" width="463" src="https://user-images.githubusercontent.com/4733304/54637283-840aa600-4a45-11e9-8e18-aa2f7a04d877.png">
> 
> implementation:
> <img alt="Screen Shot 2019-03-19 at 12 45 57 PM" width="179" src="https://user-images.githubusercontent.com/4733304/54637355-b2888100-4a45-11e9-9fca-6f1cb4bc289a.png">

**The implementation of this will be more involved, which warrants reprioritization and not delaying this one IMO. The complication is that it will require template modification for the `settings-section` component as well as a new property added to the JS side of the component which can determine which icon to show.**

**As with other things, I decided to draw the line somewhere considering the size of this PR and the need to improve the current state of the page, so there were some things that were more minor and will be follow-ups:**

- **Further brave-styles for material-styled components (arrow icons, dialog tab colors, more icons) https://github.com/brave/brave-browser/issues/3799**
- **Section Title icons https://github.com/brave/brave-browser/issues/3797**
- **Extensions Page sidebar (if we still want to do that given the usability issues) https://github.com/brave/brave-browser/issues/3771**

**I'll look at the other issues @bsclifton brought up - I wasn't experiencing some of them after yesterday's rebase, so maybe they're introduced by today's rebase.**

rossmoody commented:
> I remember there were some difficulties- but wanted to check.
> 
> _Shouldn't there be icons next to each of the titles?_
> 
> mock-up (shown in [brave/brave-browser#955](https://github.com/brave/brave-browser/issues/955) and [brave/brave-browser#964](https://github.com/brave/brave-browser/issues/964)):
> <img alt="Screen Shot 2019-03-19 at 12 48 49 PM" width="463" src="https://user-images.githubusercontent.com/4733304/54637283-840aa600-4a45-11e9-8e18-aa2f7a04d877.png">
> 
> implementation:
> <img alt="Screen Shot 2019-03-19 at 12 45 57 PM" width="179" src="https://user-images.githubusercontent.com/4733304/54637355-b2888100-4a45-11e9-9fca-6f1cb4bc289a.png">

The design mockup did have those icons in the titles but after giving it some thought, I think we should hold off on them. It's not a great UI pattern (in particular for a settings page vs a marketing page) and makes scanning the content quickly quite a bit more difficult.

petemill commented:
@rossmoody that decision seems timely! Please add the comment to https://github.com/brave/brave-browser/issues/3797 if not already

petemill commented:
@bsclifton ready for re-review.

Changes:
- Fixed: persistent scrollbar always showing on settings menu
- Fixed: autofill child items not displaying when opened (was due to move to 'Advanced' section)
- Added 'Third Party Social Media Privacy Tracker via Embeds Blocking', or whatever it's now called!
- Changed most sidebar icons
- Fixed doubled-up bottom padding in sidebar menu and overall tightened spacing and scroll logic:
  
![image](https://user-images.githubusercontent.com/741836/54730821-2a3ed480-4b48-11e9-9d55-ccc2d5851261.png)

![image](https://user-images.githubusercontent.com/741836/54730854-55c1bf00-4b48-11e9-8f7c-ed711337cff2.png)

bsclifton commented:
@petemill doh - needs a rebase

😠🤜 Chromium Seventy Fouuuurrrrrrrrrr!!1!1

rossmoody commented:
A few notes as I'm working on icons:
- If my theme is set to Dark, and I set it to light, my toolbar theme changes but my web-ui remains dark. 
- Noticed a few capitalization inconsistencies. For now let's stick with sentence case. eg. 'Social media blocking', 'Get started'
- Might be because I built debug, but I'm getting a 1-2 sec janky more icon on History panel when page loads: https://d.pr/i/qMlAf8
- Let's apply the same border to the panels that is on the sidebar: https://d.pr/i/ElCxGB

bridiver commented on browser/resources/settings/brave_settings_overrides.js:
```diff
@@ -1,8 +1,18 @@
 
-window.BravePatching.registerPolymerBehaviors({
+if (!BravePatching) {
+  console.error('BravePatching was not available to brave_settings_overrides.js')
```
when would this happen and shouldn't `BravePatching.RegisterPolymerComponentBehaviors...` be in an `else`?

petemill replied:
It's a helper error to let the developer know they haven't included `polymer_overriding.js`

bridiver commented on ui/webui/resources/polymer_overriding.js:
```diff
@@ -1,33 +1,125 @@
+        continue
+      }
+      // replace
+      chromiumIcon.innerHTML = braveIcon.innerHTML
```
same as above, shouldn't this be in an `else` to avoid an error?

petemill replied:
The `continue` on L107 avoids execution here

bridiver replied:
ah, didn't notice that at first

bridiver commented on browser/resources/settings/brave_settings_overrides.js:
```diff
@@ -1,9 +1,331 @@
 
-window.BravePatching.registerPolymerBehaviors({
+// Local Scope
+(function() {
```
does it make sense to break this file up at all? It's not terrible now, but seems like it could continue to grow and better to set a pattern of splitting into smaller files if that makes sense

petemill replied:
Yes. I was considering doing that in a follow-up with either the current html-modules approach, or invoking our webpack build target type to combine some js 'modules'. Since this PR can't land until C74 has re-landed, I may have some time to do that in this PR.

bridiver commented on browser/resources/settings/brave_settings_overrides.js:
```diff
@@ -1,9 +1,331 @@
+    } else {
+      const sectionsFromTop = document.createElement('div')
+      sectionsFromTop.innerHTML = `
+        <template is="dom-if" if="[[showPage_(pageVisibility.getStarted)]]">
```
can we load this from an html file instead of inlining?

petemill replied:
In this case, using a separate module adds a bit more complexity since all the logic for determining which page's are visible at a given time (given current url, search filtering, etc) are all in the `basic_page` module. Those properties and methods could probably be passed as attributes to another module. We can at least split this up in to a separate file (e.g `brave_settings_sections`)

petemill commented:
Waiting for C74 to land on master before merging

petemill commented:
This is ready for final C74 review. Only changes since the last review:
- Updated icon graphics
- Fixed-up some item borders (adding / removing `first` class)
- The loading 'shim' for the main WebUIs had changed in c74, so I made it work correctly with the brave-style Navigation Bar, loading in the gradient before the page content (and the bar) loads. 95678e673

petemill commented:
This is ready for final review. Only changes since the last approval:
- Rebased on final C74
- Updated icon graphics
- Fixed-up some item borders (adding / removing `first` class)
- The loading 'shim' for the main WebUIs had changed in c74, so I made it work correctly with the brave-style Navigation Bar, loading in the gradient before the page content (and the bar) loads. 95678e673

bsclifton commented:
@petemill real quick one here - I think this needs more top margin:
<img width="335" alt="Screen Shot 2019-04-10 at 1 24 33 PM" src="https://user-images.githubusercontent.com/4733304/55910890-58866180-5b94-11e9-90a4-cb3720ffb944.png">

WDYT?

rossmoody commented:
Noticed just a few remaining locale title case inconsistencies remaining from this note: https://github.com/brave/brave-core/pull/1909#issuecomment-476365444 Instead of possibly doing a confusing back and forth, how bout I submit a PR against this branch?

bsclifton commented:
@rossmoody I noticed that too- SGTM! 😄 Definitely important to be consistent

petemill commented:
@rossmoody feel free to PR or if it's just a text change add a commit to this PR

petemill commented:
@bsclifton shields description padding has been increased, and @rossmoody 's style updates integrated. Watching PR builder and attempting to fix issues...
